src/SpikeCore/SpikeCore.Web/Controllers/AuthenticationController.cs

```diff
@@ -0,0 +1,62 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Mvc;
+using SpikeCore.Data.Models;
+using SpikeCore.Web.Models;
+using SpikeCore.Web.TokenProviders;
+
+namespace SpikeCore.Web.Controllers
+{
+    /// <summary>
+    /// Provides a controller for authenticating users via validating tokens issued by the 
+    /// <see cref="PasswordlessLoginTokenProvider{TUser}"/>.
+    ///
+    /// This is based on https://www.scottbrady91.com/ASPNET-Identity/Implementing-Mediums-Passwordless-Authentication-using-ASPNET-Core-Identity.
+    /// </summary>
+    public class AuthenticationController : Controller
+    {
+        private readonly UserManager<SpikeCoreUser> _userManager;
+
+        public AuthenticationController(UserManager<SpikeCoreUser> userManager)
+        {
+            _userManager = userManager;
+        }
+
+        [HttpGet]
+        public async Task<IActionResult> TokenCallback(string email, string token)
+        {
+            var user = await _userManager.FindByEmailAsync(email);
+
+            var isValid =
+                await _userManager.VerifyUserTokenAsync(user, "PasswordlessLoginProvider", "passwordless-auth", token);
+
+            if (isValid)
+            {
+                // This prevents replay attacks by swapping out the security stamp. This works because all  tokens
+                // contain the security stamp that was used to generate them.
+                await _userManager.UpdateSecurityStampAsync(user);
+
+                await HttpContext.SignInAsync(
+                    IdentityConstants.ApplicationScheme,
+                    new ClaimsPrincipal(
+                        new ClaimsIdentity(new List<Claim> {new Claim("sub", user.Id)},
+                            IdentityConstants.ApplicationScheme)
+                    ));
```

Comment:
what about `_signInManager.SignInAsync(user)`?

Reply:
Well, I mean... that wasn't in the article I blatantly copied and pasted. But now that you mention it, that's a smarter idea... And this is the benefit of code reviews :D

Reply:
That being said, I'm wondering if they're probably supporting a more complicated environment. All we're doing here is basic role checking...

Reply:
signInManager takes care of automatically putting claims and roles associated with the user into the `ClaimsIdentity`. My guess would be that they either didn't knew or simplified it for the sake of example... or copied it from elsewhere :trollface:

Reply:
Yeah, but who'd be dumb enough to copy and paste things from the internet? Uh, asking for a friend...